astropy/coordinates/velocity_correction_funcs.py

```diff
@@ -0,0 +1,102 @@
+implementation convenience, and the public API is on SkyCoord.
+"""
+from __future__ import (absolute_import, division, print_function,
+                        unicode_literals)
+
+from .solar_system import get_body_barycentric_posvel
+from .representation import CartesianDifferential, UnitSphericalRepresentation
+from .builtin_frames import GCRS
+from .sky_coordinate import SkyCoord
+
+
+__all__ = ['radial_velocity_correction', 'helio_vector', 'bary_vector']
+
+
+def helio_vector(t, loc):
+    """
+    Computes the heliocentric velocity correction at a given time and place.
+
+    Paramters
+    ---------
+    t : astropy.time.Time
+        Time of the observation. Can be a Time array.
+    loc : astropy.coordinates.EarthLocation
+        The observer location at which to compute the correction.
+
+    Returns
+    -------
+    vh : CartesianDifferential
+        The heliocentric velocity vector
+    """
+    vsun = get_body_barycentric_posvel('sun', t)[1]
+    vearth = get_body_barycentric_posvel('earth', t)[1]
+
+    vsunearth = vearth - vsun
+
+    gcrs_p, gcrs_v = loc.get_gcrs_posvel(t)
+
+    return (vsunearth + gcrs_v).represent_as(CartesianDifferential)
+
+
+def bary_vector(t, loc):
+    """
+    Computes the Solar System barycenter velocity correction at a given time and
+    place.
+
+    Paramters
+    ---------
+    t : astropy.time.Time
+        Time of the observation. Can be a Time array.
+    loc : astropy.coordinates.EarthLocation
+        The observer location at which to compute the correction.
+
+    Returns
+    -------
+    vh : CartesianDifferential
+        The barycentric velocity vector
+    """
+    vearth = get_body_barycentric_posvel('earth', t)[1]
+
+    gcrs_p, gcrs_v = loc.get_gcrs_posvel(t)
+
+    return (vearth + gcrs_v).represent_as(CartesianDifferential)
+
+
+def radial_velocity_correction(t, loc, target, bary=True):
+    """
+    This is the private API for the SkyCoord.radial_velocity_correction method.
+    See the SkyCoord docstrings for a more detailed set of algorithm notes
+    (remember that users generally will *not* see this docstring.)
+
+    Parameters
+    ----------
+    t : astropy.time.Time
+        Time of the observation. Can be a Time array.
+    loc : astropy.coordinates.EarthLocation
+        The observer location at which to compute the correction.
+    target : astropy.coordinates.SkyCoord
+        The on-sky location at which to compute the correction.
+    bary : bool
+        If True, do barycentric correction.  If False, do heliocentric.
+
+    Returns
+    -------
+    vcorr : astropy.units.Quantity with velocity units
+        The  correction with a positive sign.  I.e., *add* this
+        to an observed radial velocity to get the heliocentric velocity.
+    """
+    if bary:
+        vcorr_cart = bary_vector(t, loc)
+    else:
+        vcorr_cart = helio_vector(t, loc)
+
+    gcrs_p, _ = loc.get_gcrs_posvel(t)
+
```

Comment:
That would seem like a bug. The underlying problem is that `obsgeoloc` is defined but not `obsgeovel` -- before trying to fix the bug, should that in fact raise an error?

For here, the solution would seem to be to not throw away the velocity information, but rather use it as well.

Possibly, `EarthLocation` should just have a `get_gcrs` method that returns this directly?